pkg/services/sqlstore/migrator/mysql_dialect.go

```diff
@@ -225,3 +225,71 @@ func (db *MySQLDialect) UpsertSQL(tableName string, keyCols, updateCols []string
 	)
 	return s
 }
+
+func (db *MySQLDialect) Lock() error {
+	sess := db.engine.NewSession()
+	defer sess.Close()
+
+	query := "SELECT GET_LOCK(?, ?)"
+	var success bool
+
+	lockName, err := db.getLockName()
+	if err != nil {
+		return fmt.Errorf("failed to generate lock name: %w", err)
+	}
+
+	// trying to obtain the lock with the specific name
+	// the lock is exclusive per session and is released explicitly by executing RELEASE_LOCK() or implicitly when the session terminates
+	// it returns 1 if the lock was obtained successfully,
+	// 0 if the attempt timed out (for example, because another client has previously locked the name),
+	// or NULL if an error occurred
+	// starting from MySQL 5.7 it is even possible for a given session to acquire multiple locks for the same name
+	// however other sessions cannot acquire a lock with that name until the acquiring session releases all its locks for the name.
+	_, err = sess.SQL(query, lockName, 0).Get(&success)
```

Comment:
I have set timeout to 0 so it fail fast and it won't wait at all if the database is already locked.

Reply:
This means that the Grafana instance not acquiring the lock will continue to init and start Grafana, right? Have you considered this a problem?

Reply:
no, AFAIU if it's not able to acquire the lock returns an error and falls in [this](https://github.com/grafana/grafana/blob/53fbb825f162b6044d51a408b7ea7edd236ef1d4/pkg/services/sqlstore/migrator/migrator.go#L102) case, so the server will fail to start.

Reply:
Great, that feels like a viable solution :+1: In Kubernetes this would mean that new pods will be started until migrations are complete and the `ReplicaSet` is fulfilled with number of running/health pods.

Reply:
I think we want to make the timeout configurable and allow the Grafana instance to retry the migrations after a certain time. This will work in K8S but we also alert on pod restarting too many times to catch configuration errors and other unexpected behaviors. 

In the on-prem scenario I think retries and configurable timeout would be very appropriated since on-prem tooling might not solve the instance life cycle as good as K8s


Making the timeout configurable should be good enough to get this merged and retries could be solve in a follow up PR.

Reply:
I agree, it would be nice to be able to set this timeout in HG. I think it would be preferrable for a 2nd instance to wait a few seconds to get a lock, then perform a noop on migration (as they're already completed, say), compared to crashLoop until the lock becomes available.